src/uct/ib/base/ib_md.c

```diff
@@ -815,36 +825,85 @@ static void uct_ib_mem_init(uct_ib_mem_t *memh, unsigned uct_flags,
     if (uct_flags & UCT_MD_MEM_ACCESS_REMOTE_ATOMIC) {
         memh->flags |= UCT_IB_MEM_ACCESS_REMOTE_ATOMIC;
     }
+
+    if (uct_flags & UCT_MD_MEM_FLAG_ON_DEVICE) {
+        memh->flags |= UCT_IB_MEM_FLAG_DM;
+    }
 }
 
+#if HAVE_IBV_EXP_DM
+static struct ibv_exp_dm *uct_ib_mem_alloc_dm(uct_ib_md_t *md, size_t length)
+{
+    struct ibv_exp_alloc_dm_attr dm_attr = {.length = length};
+
+    return UCS_PROFILE_CALL(ibv_exp_alloc_dm, md->dev.ibv_context, &dm_attr);
+}
+#endif
+
 static ucs_status_t uct_ib_mem_alloc(uct_md_h uct_md, size_t *length_p,
                                      void **address_p, unsigned flags,
                                      const char *alloc_name, uct_mem_h *memh_p)
 {
 #if HAVE_DECL_IBV_EXP_ACCESS_ALLOCATE_MR
-    uct_ib_md_t *md = ucs_derived_of(uct_md, uct_ib_md_t);
+    uct_ib_md_t *md    = ucs_derived_of(uct_md, uct_ib_md_t);
+    uint32_t comp_mask = 0;
     ucs_status_t status;
     uint64_t exp_access;
     uct_ib_mem_t *memh;
     size_t length;
 
+#if HAVE_IBV_EXP_DM
+    if (!(flags & UCT_MD_MEM_FLAG_ON_DEVICE) &&
+        !md->config.enable_contig_pages) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+    if ((flags & UCT_MD_MEM_FLAG_ON_DEVICE) &&
+        (flags & UCT_MD_MEM_FLAG_FIXED)) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+#else
     if (!md->config.enable_contig_pages) {
         return UCS_ERR_UNSUPPORTED;
     }
+    if (flags & UCT_MD_MEM_FLAG_ON_DEVICE) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+#endif
 
     memh = uct_ib_memh_alloc(md);
     if (memh == NULL) {
         status = UCS_ERR_NO_MEMORY;
         goto err;
     }
 
-    length     = *length_p;
+    length = *length_p;
+
+#if HAVE_IBV_EXP_DM
+    if (flags & UCT_MD_MEM_FLAG_ON_DEVICE) {
+        memh->dm = uct_ib_mem_alloc_dm(md, length);
+
+        if (memh->dm == NULL) {
+            flags &= ~UCT_MD_MEM_FLAG_ON_DEVICE;
+        } else {
+            comp_mask = IBV_EXP_REG_MR_DM;
+        }
+    }
+#endif
+
     exp_access = uct_ib_md_access_flags(md, flags, length) |
-                 IBV_EXP_ACCESS_ALLOCATE_MR;
-    status = uct_ib_md_reg_mr(md, NULL, length, exp_access, 0, &memh->mr);
+                 !(flags & UCT_MD_MEM_FLAG_ON_DEVICE) ?
+                 IBV_EXP_ACCESS_ALLOCATE_MR : UCT_IB_MEM_ACCESS_FLAGS;
+
+    status = uct_ib_md_reg_mr(md, NULL, length, exp_access, comp_mask, memh->dm, 0, &memh->mr);
```

Comment:
no fallback in case MEMIC is not available, especially to the buggy and potentially disabled contig pages allocator. need to return UCS_ERR_NO_RESOURCES, and fallback in uct_alloc_mem